src/mod_cache_users.erl

```diff
@@ -0,0 +1,236 @@
+    ].
+
+-spec does_user_exist(HostType :: mongooseim:host_type(), JID :: jid:jid()) -> boolean().
+does_user_exist(HostType, #jid{luser = LUser, lserver = LServer} = JID) ->
+    case does_cached_user_exist(HostType, LServer, LUser) of
+        true -> true;
+        false ->
+            case does_stored_user_exist(HostType, JID) of
+                true ->
+                    put_user_into_cache(HostType, LServer, LUser),
+                    true;
+                false -> false
+            end
+    end.
+
+%%====================================================================
+%% Hooks
+%%====================================================================
+
+-spec remove_user(Acc :: mongoose_acc:t(),
+                  LUser :: jid:luser(),
+                  LServer :: jid:lserver() | string()) -> mongoose_acc:t().
+remove_user(Acc, LUser, LServer) ->
+    HostType = mongoose_acc:host_type(Acc),
+    Key = key(LUser, LServer),
+    Tab = tbl_name(HostType),
+    del_member(Tab, Key, ets:first(Tab)),
+    Acc.
+
+-spec remove_domain(mongoose_hooks:simple_acc(),
+                    mongooseim:host_type(), jid:lserver()) ->
+    mongoose_hooks:simple_acc().
+remove_domain(Acc, HostType, Domain) ->
+    Tab = tbl_name(HostType),
+    del_domain(Tab, Domain, ets:first(Tab)),
+    Acc.
+
+%%====================================================================
+%% Helpers
+%%====================================================================
+
+-spec does_stored_user_exist(mongooseim:host_type(), jid:jid()) -> boolean().
+does_stored_user_exist(HostType, JID) ->
+    true =:= ejabberd_auth:does_stored_user_exist(HostType, JID).
+
+-spec does_cached_user_exist(mongooseim:host_type(), jid:lserver(), jid:luser()) -> boolean().
+does_cached_user_exist(HostType, LServer, LUser) ->
+    Key = key(LUser, LServer),
+    Tab = tbl_name(HostType),
+    is_member(Tab, Key, ets:last(Tab)).
+
+-spec put_user_into_cache(mongooseim:host_type(), jid:lserver(), jid:luser()) -> ok.
+put_user_into_cache(HostType, LServer, LUser) ->
+    Key = key(LUser, LServer),
+    Tab = tbl_name(HostType),
+    put_member(Tab, Key, ets:last(Tab)).
+
+-spec key(jid:luser(), jid:lserver()) -> {jid:lserver(), jid:luser()}.
+key(LUser, LServer) ->
+    {LServer, LUser}.
+
+is_member(_, _, '$end_of_table') ->
+    false;
+is_member(Tab, Key, Table) ->
+    case ets:lookup(Tab, Table) of
+        [{_, EtsSegment}] ->
+            case ets:member(EtsSegment, Key) of
+                false -> is_member(Tab, Key, ets:prev(Tab, Table));
+                true -> true
+            end;
+        [] ->
+           is_member(Tab, Key, ets:prev(Tab, Table))
+    end.
+
+%% There's a chance that by the time we insert in the ets table, this table is not
+%% the first anymore because the cleaner has taken action and pushed it behind.
+%% That's fine, worst case this record will live a segment less than expected.
+put_member(Tab, Key, Table) ->
```

Comment:
Tab, Table... Poor people trying to read that later XDD
ParentTab, SegmentTab (ends with Tab - we know its a tab).